src/pnpq/devices/switch_thorlabs_osw1310e.py

```diff
@@ -1,48 +1,185 @@
-    ):
-        self.conn = Serial()
-        self.conn.baudrate = 115200
-        self.conn.bytesize = 8
-        self.conn.parity = "N"
-        self.conn.rtscts = True
-
-        self.port = serial_port
-        self.conn.port = self.port
-        self.device_sn = serial_number
-
-        find_port = False
-        if self.device_sn is not None:
-            available_ports = serial.tools.list_ports.comports()
-            for ports in available_ports:
-                if ports.serial_number == self.device_sn:
-                    self.conn.port = ports.device
-                    find_port = True
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        self.close()
+
+    def open(self) -> None:
+        # These devices tend to take a few seconds to start up, and
+        # this library tends to be used as part of services that start
+        # automatically on computer boot. For safety, wait here before
+        # continuing initialization.
+        time.sleep(1)
+
+        port_found = False
+        port = None
+        for possible_port in serial.tools.list_ports.comports():
+            if possible_port.serial_number == self.serial_number:
+                port = possible_port
+                port_found = True
+                break
+        if not port_found:
+            raise ValueError(
+                f"Serial number {self.serial_number} could not be found, failing intialization."
+            )
+        assert port is not None
+
+        # Initializing the connection by passing a port to the Serial
+        # constructor immediately opens the connection. It is not
+        # necessary to call open() separately.
+
+        object.__setattr__(
+            self,
+            "connection",
+            Serial(
+                baudrate=self.baudrate,
+                bytesize=self.bytesize,
+                exclusive=self.exclusive,
+                parity=self.parity,
+                port=port.device,
+                rtscts=self.rtscts,
+                stopbits=self.stopbits,
+                timeout=self.timeout,
+            ),
+        )
+
+        time.sleep(0.1)
+        self.connection.flush()
+
+        # Remove anything that might be left over in the buffer from
+        # previous runs
+        self.connection.reset_input_buffer()
+        self.connection.reset_output_buffer()
+
+    def close(self) -> None:
+        with self.mutex_lock:
+            if self.connection.is_open:
+                self.connection.flush()
+                self.connection.close()
+
+    def set_state(self, state: State) -> None:
+        with self.mutex_lock, timeout(3) as check_timeout:
+            while check_timeout():
+                if state == State.BAR:
+                    self.connection.write(SET_STATE_BAR_COMMAND)
+                elif state == State.CROSS:
+                    self.connection.write(SET_STATE_CROSS_COMMAND)
```

Comment:
I don't think this is what you want to do. You want to write the command once, then check for its status. You don't want to keep writing the command over and over.